File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.py
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if not head or left == right:
            return head

        dummy = ListNode(0)
        dummy.next = head
        prev = dummy

        # Move prev to the node just before 'left'
        for _ in range(left - 1):
            prev = prev.next


        # Reverse the sublist from left to right
        curr = prev.next
        prev_sublist = None

        for _ in range(right - left + 1):
            next_node = curr.next
            curr.next = prev_sublist
            prev_sublist = curr
            curr = next_node
        print(prev_sublist)
        print(curr)
        print(prev)
        # Connect the reversed sublist back to the original list
        
        prev.next.next = curr
        print(prev)
        prev.next = prev_sublist
        print(prev)
        
        return dummy.next

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummy = new ListNode();
        ListNode prev = dummy;
        prev.next = head;

        for (int i = 0; i < left - 1; i++) {
            prev = prev.next;
        }

        ListNode prevSubList = prev;
        ListNode leftNode = prevSubList.next;

        for (int i = 0; i < (right -left); i++) {
            prev = prev.next;
        }

        ListNode rightNode = prev.next;
        ListNode afterSubList = rightNode.next;

        prevSubList.next = null;
        rightNode.next = null;

        ListNode curr = leftNode;
        prev = null;
        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        prevSubList.next = rightNode;
        leftNode.next = afterSubList;

        return dummy.next;
    }
}